### PR TITLE
ci: making it easier to upgrade node

### PR DESCRIPTION
**Pull request by lodmfjord**
# Making it easier to update node

This makes it so it is easier to upgrade node in the future.

## What

* Defining node version in one place (instead of many)
* * github workflows uses package.json as a source of node version
* * get-node-version.mjs find the correct image tag for docker
* Cache key now adds arch, node version and yarn version
* * node_modules built on different arch would break (this is currently not a problem)
* * node_modules built on different major version of node would break (this problem while upgrading)

## Why

Upgrading node has been a problem in the past. This is not something that should be done often but should be easy to test.

## Summary by CodeRabbit

- **New Features**
  - Enhanced Docker builds with dynamic Node.js version handling for increased flexibility and efficiency.
  - Added scripts for calculating node module hashes and fetching the latest Node.js Docker image tags.

- **Bug Fixes**
  - Improved code revision logic in CI workflows for more precise builds and tests.

- **Chores**
  - Updated GitHub Actions and workflow configurations to utilize the latest setup-node version and dependencies.
  - Optimized caching mechanisms for node modules and generated files to enhance CI performance.

- **Refactor**
  - Restructured CI scripts and workflows to improve readability and maintainability.

### CODE REVIEW

File: .github/workflows/pullrequest.yml [hunk 1]
```diff
@@ -80,11 +80,7 @@ jobs:
       - name: Calculate cache key for node modules
         id: calculate_node_modules_hash
         run: |
-          PACKAGE_JSON_HASH=$(cat package.json | jq '{resolutions,dependencies,devDependencies}' | sha1sum -t | cut -f1 -d" ")
-          echo "PACKAGE_JSON_HASH: $PACKAGE_JSON_HASH"
-          export NODE_MODULES_HASH=${{ runner.os }}-${{ hashFiles('yarn.lock') }}-$PACKAGE_JSON_HASH
-          echo "NODE_MODULES_HASH: $NODE_MODULES_HASH"
-          echo "node-modules-hash=$NODE_MODULES_HASH" >> $GITHUB_OUTPUT
+          echo "node-modules-hash=$(./scripts/ci/get-node-modules-hash.mjs)" >> $GITHUB_OUTPUT
```
Comment (busla) [inline .github/workflows/pullrequest.yml:83]: could you add a debug below so we see the hash?

```suggestion
          HASH="$(./scripts/ci/get-node-modules-hash.mjs)"
          echo "node-modules-hash: ${HASH}"
          echo "node-modules-hash=${HASH}" >> $GITHUB_OUTPUT
```

File: .github/workflows/push.yml [hunk 2]
```diff
@@ -72,6 +73,7 @@ jobs:
             # Avoid too long resource names
             GIT_BRANCH_DEPLOY=${GIT_BRANCH_DEPLOY:0:50}
           fi
+          echo "NODE_IMAGE_TAG=$(./scripts/ci/get-node-version.mjs)" >> $GITHUB_OUTPUT
```
Comment (busla) [inline .github/workflows/push.yml:76]:
```suggestion
          NODE_IMAGE_TAG="$(./scripts/ci/get-node-version.mjs)"
          echo "NODE_IMAGE_TAG: ${NODE_IMAGE_TAG}"
          echo "NODE_IMAGE_TAG=${NODE_IMAGE_TAG}" >> $GITHUB_OUTPUT
```

File: .github/workflows/push.yml [hunk 3]
```diff
@@ -277,11 +279,7 @@ jobs:
       - name: Calculate cache key for node modules
         id: calculate_node_modules_hash
         run: |
-          PACKAGE_JSON_HASH=$(cat package.json | jq '{resolutions,dependencies,devDependencies}' | sha1sum -t | cut -f1 -d" ")
-          echo "PACKAGE_JSON_HASH: $PACKAGE_JSON_HASH"
-          export NODE_MODULES_HASH=${{ runner.os }}-${{ hashFiles('yarn.lock') }}-$PACKAGE_JSON_HASH
-          echo "NODE_MODULES_HASH: $NODE_MODULES_HASH"
-          echo "node-modules-hash=$NODE_MODULES_HASH" >> $GITHUB_OUTPUT
+          echo "node-modules-hash=$(./scripts/ci/get-node-modules-hash.mjs)" >> $GITHUB_OUTPUT
```
Comment (busla) [inline .github/workflows/push.yml:282]:
```suggestion
          HASH="$(./scripts/ci/get-node-modules-hash.mjs)"
          echo "node-modules-hash: ${HASH}"
          echo "node-modules-hash=${HASH}" >> $GITHUB_OUTPUT
```

File: .github/workflows/push.yml [hunk 4]
```diff
@@ -601,15 +596,19 @@ jobs:
         continue-on-error: true
         id: dockerbuild
         if: steps.gather.outcome == 'success'
-        run: ./scripts/ci/run-in-parallel.sh 90_$DOCKER_TYPE
         env:
-          EXTRA_DOCKER_BUILD_ARGS: '--build-arg DOCKER_IMAGE_REGISTRY=${{ env.DOCKER_BASE_IMAGE_REGISTRY }} --build-arg GIT_SHA=${{ github.sha }}'
+          NODE_IMAGE_TAG: ${{ needs.pre-checks.outputs.NODE_IMAGE_TAG }}
+        run: |
+          EXTRA_DOCKER_BUILD_ARGS='--build-arg DOCKER_IMAGE_REGISTRY=${{ env.DOCKER_BASE_IMAGE_REGISTRY }} --build-arg GIT_SHA=${{ github.sha }} --build-arg NODE_IMAGE_TAG=$NODE_IMAGE_TAG'
```
Comment (busla) [inline .github/workflows/push.yml:602]: `EXTRA_DOCKER_BUILD_ARGS` is wrapped in single quotes so confirm that `$NODE_IMAGE_TAG` expands correctly.

File: .github/workflows/push.yml [hunk 5]
```diff
@@ -601,15 +596,19 @@ jobs:
...
+          ./scripts/ci/run-in-parallel.sh 90_$DOCKER_TYPE
 
       - name: Building Docker images Retry #This only exists until GHA starts supporting this
         if: steps.gather.outcome == 'success' && steps.dockerbuild.outcome == 'failure'
-        run: ./scripts/ci/run-in-parallel.sh 90_$DOCKER_TYPE
         env:
-          EXTRA_DOCKER_BUILD_ARGS: '--build-arg DOCKER_IMAGE_REGISTRY=${{ env.DOCKER_BASE_IMAGE_REGISTRY }} --build-arg GIT_SHA=${{ github.sha }}'
+          NODE_IMAGE_TAG: ${{ needs.pre-checks.outputs.NODE_IMAGE_TAG }}
+        run: |
+          EXTRA_DOCKER_BUILD_ARGS='--build-arg DOCKER_IMAGE_REGISTRY=${{ env.DOCKER_BASE_IMAGE_REGISTRY }} --build-arg GIT_SHA=${{ github.sha }} --build-arg NODE_IMAGE_TAG=$NODE_IMAGE_TAG'
```
Comment (busla) [inline .github/workflows/push.yml:610]: same here, check `$NODE_IMAGE_TAG` expansion.

File: .github/workflows/push.yml [hunk 6]
```diff
@@ -644,13 +643,14 @@ jobs:
         working-directory: infra
         run: |
           ./scripts/build-docker-container.sh $DOCKER_TAG
+          EXTRA_DOCKER_BUILD_ARGS="--build-arg DOCKER_IMAGE_REGISTRY=${{ env.DOCKER_BASE_IMAGE_REGISTRY }} --build-arg NODE_IMAGE_TAG=$NODE_IMAGE_TAG"
```
Comment (busla) [inline .github/workflows/push.yml:646]: confirm expansion

File: infra/scripts/Dockerfile [hunk 7]
```diff
@@ -1,9 +1,10 @@
 ARG DOCKER_IMAGE_REGISTRY=public.ecr.aws
-FROM $DOCKER_IMAGE_REGISTRY/docker/library/node:18-alpine3.15 as runner
+ARG NODE_IMAGE_TAG
+FROM $DOCKER_IMAGE_REGISTRY/docker/library/node:$NODE_IMAGE_TAG as runner
```
Comment (busla) [inline infra/scripts/Dockerfile:3]:
```suggestion
FROM ${DOCKER_IMAGE_REGISTRY}/docker/library/node:${NODE_IMAGE_TAG} as runner
```

File: infra/scripts/Dockerfile [hunk 8]
```diff
@@ -1,9 +1,10 @@
...
 
 RUN apk add postgresql-client bash
 
-FROM $DOCKER_IMAGE_REGISTRY/docker/library/node:18-alpine3.15 as build
+FROM $DOCKER_IMAGE_REGISTRY/docker/library/node:$NODE_IMAGE_TAG as build
```
Comment (busla) [inline infra/scripts/Dockerfile:7]:
```suggestion
FROM ${DOCKER_IMAGE_REGISTRY}/docker/library/node:${NODE_IMAGE_TAG} as build
```

File: scripts/ci/10_prepare-docker-deps.sh [hunk 9]
```diff
@@ -7,6 +7,11 @@ DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
 source "$DIR"/_common.sh
 
 mkdir -p "$PROJECT_ROOT"/cache
+
+NODE_IMAGE_TAG=${NODE_IMAGE_TAG:-$(./scripts/ci/get-node-version.mjs)}
+
+echo $NODE_IMAGE_TAG
```
Comment (busla) [inline scripts/ci/10_prepare-docker-deps.sh]:
```suggestion
echo "NODE_IMAGE_TAG: ${NODE_IMAGE_TAG}"
```

File: scripts/ci/Dockerfile [hunk 10]
```diff
@@ -1,7 +1,8 @@
 # This is a multi-stage Dockerfile which contains all CI-related operations as well as images to be deployed in production
 ARG PLAYWRIGHT_VERSION
 ARG DOCKER_IMAGE_REGISTRY=public.ecr.aws
-FROM $DOCKER_IMAGE_REGISTRY/docker/library/node:18-alpine3.15 as deps
+ARG NODE_IMAGE_TAG
+FROM $DOCKER_IMAGE_REGISTRY/docker/library/node:$NODE_IMAGE_TAG as deps
```
Comment (busla) [inline scripts/ci/Dockerfile:5]:
```suggestion
FROM ${DOCKER_IMAGE_REGISTRY}/docker/library/node:${NODE_IMAGE_TAG} as deps
```

File: scripts/ci/Dockerfile [hunk 11]
```diff
@@ -32,7 +33,7 @@ ENV NODE_OPTIONS="--max-old-space-size=8192"
 
 RUN yarn run build ${APP} --prod
 
-FROM $DOCKER_IMAGE_REGISTRY/docker/library/node:18-alpine3.15 as output-base
+FROM $DOCKER_IMAGE_REGISTRY/docker/library/node:$NODE_IMAGE_TAG as output-base
```
Comment (busla) [inline scripts/ci/Dockerfile:36]:
```suggestion
FROM ${DOCKER_IMAGE_REGISTRY}/docker/library/node:${NODE_IMAGE_TAG} as output-base
```

File: scripts/ci/_common.mjs [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import { readFile } from 'node:fs/promises'
+import { dirname, resolve } from 'path'
+import { fileURLToPath } from 'url'
+
+const __dirname = dirname(fileURLToPath(import.meta.url))
+export const ROOT = resolve(__dirname, '..', '..')
+
+export async function getPackageJSON(filePath = resolve(ROOT, 'package.json')) {
```
Comment (busla) [inline scripts/ci/_common.mjs:8]: es6?

File: scripts/ci/get-node-modules-hash.mjs [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+import { arch, platform } from 'os'
+import crypto from 'node:crypto'
+
+import { ROOT, getPackageJSON } from './_common.mjs'
+
+process.stdout.write(
+  `${getPlatformString()}-${await getYarnLockHash()}-${await getPackageHash()}-${await getNodeVersionString()}`,
+)
+
+async function getNodeVersionString() {
```
Comment (busla) [inline scripts/ci/get-node-modules-hash.mjs:13]: es6?

File: scripts/ci/get-node-version.mjs [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+async function getVersion(
+  version,
+  withAlpine = true,
+  architecture = 'amd64',
+  url = null,
+) {
+  try {
+    const baseURL =
+      url ??
+      'https://hub.docker.com/v2/repositories/library/node/tags?page_size=100'
```
Comment (busla) [inline scripts/ci/get-node-version.mjs:24]: can we move this to a constant at the top of the file?

**Review by busla — CHANGES_REQUESTED** [review]
LGTM, few suggestions

File: scripts/ci/get-node-version.mjs (hunk 14 above)
Comment (lodmfjord) [reply scripts/ci/get-node-version.mjs:24]: Cool. Will do.

File: scripts/ci/get-node-modules-hash.mjs (hunk 13 above)
Comment (lodmfjord) [reply scripts/ci/get-node-modules-hash.mjs:13]: Yeah esm allows await in top :)

**Comment by lodmfjord** [comment]
@coderabbitai resolve

File: scripts/ci/get-node-modules-hash.mjs [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    const yarnVersion = content?.engines?.yarn;
+    if (!nodeVersion) {
+        throw new Error('Node version not defined');
+    }
+    if (!yarnVersion) {
+        throw new Error('Yarn version not defined');
+    }
+
+    return `${nodeVersion}-${yarnVersion}`;
+}
```
[the remark this answers is not in the document]
Comment (lodmfjord) [reply scripts/ci/get-node-modules-hash.mjs]: @coderabbitai resolve

File: scripts/ci/get-node-modules-hash.mjs [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+async function getPackageHash(keys = ["resolutions", "dependencies", "devDependencies"]) {
+    const content = await getPackageJSON()
+    const value = keys.reduce((a, b) => {
+        return {
+            ...a,
+            [b]: content[b]
+        }
+    }, {});
+    return crypto.createHash('sha256').update(JSON.stringify(value)).digest('hex');
+}
```
[the remark this answers is not in the document]
Comment (lodmfjord) [reply scripts/ci/get-node-modules-hash.mjs]: @coderabbitai resolve

File: package.json [hunk 17]
```diff
@@ -5,7 +5,8 @@
   "engines": {
     "npm": "please-use-yarn",
     "node": "18.16.0",
-    "yarn": "3.2"
+    "yarn": "3.2",
+    "all-services-are-affected": true
```
Comment (saevarma) [inline package.json:9]: Only PR trick to make everything affected?

Comment (lodmfjord) [reply package.json:9]: Yeah. Sorry. This won't be merged.

**Review by saevarma — APPROVED** [review]
The single core file LGTM

**Comment by andes-it** [comment]
This PR's had a 15-day siesta. Lost its spark, did it? It's making an exit in 5 days unless the stale label disappears.

**Comment by lodmfjord** [comment]
@coderabbitai ignore

File: scripts/ci/get-node-version.mjs [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+  'https://hub.docker.com/v2/repositories/library/node/tags?page_size=100'
+
+const nodeVersion = await getPackageVersion()
+const version = await getVersion(nodeVersion)
+
+if (!version) {
+  process.stderr.write(`Failed getting docker image for ${nodeVersion}`)
+  process.exit(1)
+}
+process.stdout.write(version)
```
Comment (AndesKrrrrrrrrrrr) [inline scripts/ci/get-node-version.mjs:14]:
```suggestion
console.error(version)
```

File: scripts/ci/get-node-version.mjs [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+
+const DOCKERHUB_BASE_URL =
+  'https://hub.docker.com/v2/repositories/library/node/tags?page_size=100'
+
+const nodeVersion = await getPackageVersion()
+const version = await getVersion(nodeVersion)
+
+if (!version) {
+  process.stderr.write(`Failed getting docker image for ${nodeVersion}`)
+  process.exit(1)
```
Comment (AndesKrrrrrrrrrrr) [inline scripts/ci/get-node-version.mjs:12]: Consider moving to a `main` function which can be `return`ed from instead

File: scripts/ci/Dockerfile [hunk 20]
```diff
@@ -1,7 +1,8 @@
 # This is a multi-stage Dockerfile which contains all CI-related operations as well as images to be deployed in production
 ARG PLAYWRIGHT_VERSION
 ARG DOCKER_IMAGE_REGISTRY=public.ecr.aws
-FROM $DOCKER_IMAGE_REGISTRY/docker/library/node:18-alpine3.15 as deps
+ARG NODE_IMAGE_TAG
```
Comment (AndesKrrrrrrrrrrr) [inline scripts/ci/Dockerfile:4]: Please add _any_ sane default

File: scripts/ci/get-node-version.mjs (hunk 18 above)
Comment (lodmfjord) [reply scripts/ci/get-node-version.mjs:14]: No this is an output that should go to stdout

**Pull request merged**